### PR TITLE
Abstract chaincodes into CChainCode

### PR DESCRIPTION
**Pull request by jonasschnelli**
Rehashed work from sipa. Slightly improves current Bip32 support (CExtKey, CExtPubKey).
**Code is unused at the moment (only used from `bip32_tests.cpp`)**

### CODE REVIEW

File: src/key.h [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #ifndef BITCOIN_KEY_H
 #define BITCOIN_KEY_H
 
+#include "pubkey.h"
```
Comment (jonasschnelli) [inline src/key.h]: I hope nobody complains about this new include?

**Comment by sipa** [comment]
utACK

File: src/key.h (hunk 1 above)
Comment (theuni) [inline src/key.h]: From my POV, that's on the list of "safe to include everywhere" since it's used in libbitcoinconsensus.

**Comment by theuni** [comment]
Looks good to me, but if you're going to the trouble of abstracting CChainCode, you may as well treat chaincodes as objects wholesale. See https://github.com/theuni/bitcoin/commit/58fe156ec2672b354650c3e3d89d0724723bbc34 for a quick (untested) example.

**Comment by jonasschnelli** [comment]
Thanks @theuni! Right. This would be the next logical step. Reviewed and added (as it is) your https://github.com/theuni/bitcoin/commit/58fe156ec2672b354650c3e3d89d0724723bbc34 to this PR.

**Comment by sipa** [comment]
key already implicitly depends on pubkey anyway (through the forward
declaration).

**Comment by sipa** [comment]
utACK both commits

**Comment by jgarzik** [comment]
ACK

**Comment by laanwj** [comment]
Playing devil's advocate: how is  a `CChainCode` different from a `uint256`? (the new type without arithmethic operators) It has a superset of the methods (like `SetNull` and such) and serializes the same way.

**Comment by theuni** [comment]
@laanwj heh, very good point.

**Comment by jonasschnelli** [comment]
Indeed!
If there are no concerns i'll change this so it will use `uint256` instead of `CChainCode` (or the current `unsigned char chainCode[32]`)

**Comment by sipa** [comment]
If you plan to later extend CChainCode with extra functionality, you could at this point just do a

typedef uint256 ChainCode;

(leave the C prefix of for new classes, we don't care)

**Comment by jonasschnelli** [comment]
Replaced the CChainCode struct with a uint256 typedef.

File: src/hash.cpp [hunk 2]
```diff
@@ -71,14 +72,14 @@ unsigned int MurmurHash3(unsigned int nHashSeed, const std::vector<unsigned char
...
+void BIP32Hash(const ChainCode &chainCode, unsigned int nChild, unsigned char header, const unsigned char data[32], unsigned char output[64])
 {
     unsigned char num[4];
     num[0] = (nChild >> 24) & 0xFF;
     num[1] = (nChild >> 16) & 0xFF;
     num[2] = (nChild >>  8) & 0xFF;
     num[3] = (nChild >>  0) & 0xFF;
-    CHMAC_SHA512(chainCode, 32).Write(&header, 1)
+    CHMAC_SHA512(chainCode.begin(), chainCode.size()).Write(&header, 1)
                                .Write(data, 32)
```
Comment (sipa) [inline src/hash.cpp]: Indentation is off now. No problem if you don't want to keep this style, but now it becomes inconsistent.

Comment (jonasschnelli) [inline src/hash.cpp]: Agreed. Removed the indentation.

**Comment by laanwj** [comment]
utACK

**Pull request merged**